File: packages/shield/src/risk.js
import generalisations from './data/generalisations.json' with { type: 'json' };

/**
 * Analyse re-identification risk of a dataset using k-anonymity and
 * rare-category detection.
 *
 * @param {Array<Array<*>>} data - Array of row arrays (no header row).
 * @param {object} config
 * @param {number[]} config.quasiIdentifiers - Column indices for k-anonymity grouping.
 * @param {Array<{index: number, name: string}>} [config.categoricalColumns] - Columns to check for rare categories.
 * @param {number|'auto'} [config.kThreshold='auto'] - Fixed k or 'auto' (k=5 for <20 rows, k=3 for >=20).
 * @returns {{ kAnonymity: object, rareCategories: Array, recommendations: string[] }}
 */
export function analyseRisk(data, config) {
  const {
    quasiIdentifiers = [],
    categoricalColumns = [],
    kThreshold = 'auto',
  } = config || {};

  const totalStudents = data.length;

  const k =
    kThreshold === 'auto'
      ? totalStudents < 20
        ? 5
        : 3
      : kThreshold;

  const kAnonymity = computeKAnonymity(data, quasiIdentifiers, k, totalStudents);
  const rareCategories = detectRareCategories(data, categoricalColumns, k);
  const recommendations = buildRecommendations(kAnonymity, rareCategories, k);

  return { kAnonymity, rareCategories, recommendations };
}

/**
 * Group rows by quasi-identifier values and flag small groups.
 *
 * @param {Array<Array<*>>} data
 * @param {number[]} qiCols
 * @param {number} k
 * @param {number} totalStudents
 * @returns {{ threshold: number, totalStudents: number, safeStudents: number, flaggedGroups: Array }}
 */
function computeKAnonymity(data, qiCols, k, totalStudents) {
  if (totalStudents === 0 || qiCols.length === 0) {
    return {
      threshold: k,
      totalStudents,
      safeStudents: totalStudents,
      flaggedGroups: [],
    };
  }

  // Build groups keyed by composite quasi-identifier string.
  const groups = new Map();

  for (let i = 0; i < data.length; i++) {
    const row = data[i];
    const key = qiCols.map((col) => String(row[col] ?? '')).join('\x00');
    if (!groups.has(key)) {
      groups.set(key, { attrs: qiCols.map((col) => row[col]), indices: [] });
    }
    groups.get(key).indices.push(i);
  }

  const flaggedGroups = [];
  let flaggedCount = 0;

  for (const [, group] of groups) {
    if (group.indices.length < k) {
      const attributes = {};
      qiCols.forEach((col, idx) => {
        attributes[`col${col}`] = group.attrs[idx];
      });
      flaggedGroups.push({
        attributes,
        count: group.indices.length,
        rowIndices: group.indices,
      });
      flaggedCount += group.indices.length;
    }
  }

  // Sort flagged groups by count ascending for deterministic output.
  flaggedGroups.sort((a, b) => a.count - b.count);

  return {
    threshold: k,
    totalStudents,
    safeStudents: totalStudents - flaggedCount,
    flaggedGroups,
  };
}

/**
 * Detect rare categories in categorical columns and suggest generalisations.
 *
 * @param {Array<Array<*>>} data
 * @param {Array<{index: number, name: string}>} categoricalColumns
 * @param {number} k
 * @returns {Array<{column: number, columnName: string, value: *, count: number, suggestion: string|null}>}
 */
function detectRareCategories(data, categoricalColumns, k) {
  if (!categoricalColumns || categoricalColumns.length === 0 || data.length === 0) {
    return [];
  }

  const rareThreshold = Math.min(3, k);
  const results = [];

  for (const { index, name } of categoricalColumns) {
    // Count occurrences.
    const counts = new Map();
    for (const row of data) {
      const val = row[index] ?? '';
      counts.set(val, (counts.get(val) || 0) + 1);
    }

    for (const [value, count] of counts) {
      if (count < rareThreshold) {
        const suggestion = lookupGeneralisation(name, value);
        results.push({
          column: index,
          columnName: name,
          value,
          count,
          suggestion,
        });
      }
    }
  }

  // Sort by count ascending, then by column index for determinism.
  results.sort((a, b) => a.count - b.count || a.column - b.column);

  return results;
}

/**
 * Look up a generalisation suggestion for a given category name and value.
 *
 * @param {string} categoryName - Column / category name (e.g. "SEN", "PP").
 * @param {*} value - The rare value to look up.
 * @returns {string|null}
 */
function lookupGeneralisation(categoryName, value) {
  const strValue = String(value);

  // Try exact category match first (case-insensitive key search).
  for (const [catKey, mapping] of Object.entries(generalisations)) {
    if (catKey.toLowerCase() === categoryName.toLowerCase()) {
      for (const [mapKey, suggestion] of Object.entries(mapping)) {
        if (mapKey.toLowerCase() === strValue.toLowerCase()) {
          return suggestion;
        }
      }
      // Category found but value not in mapping.
      return null;
    }
  }

  // Category not found — search all categories for the value.
  for (const [, mapping] of Object.entries(generalisations)) {
    for (const [mapKey, suggestion] of Object.entries(mapping)) {
      if (mapKey.toLowerCase() === strValue.toLowerCase()) {
        return suggestion;
      }
    }
  }

  return null;
}

/**
 * Build plain-English recommendation strings.
 *
 * @param {{ threshold: number, totalStudents: number, safeStudents: number, flaggedGroups: Array }} kAnonymity
 * @param {Array} rareCategories
 * @param {number} k
 * @returns {string[]}
 */
function buildRecommendations(kAnonymity, rareCategories, k) {
  const recommendations = [];
  const { totalStudents, safeStudents, flaggedGroups } = kAnonymity;

  if (totalStudents === 0) {
    recommendations.push('No data to analyse.');
    return recommendations;
  }

  const flaggedCount = totalStudents - safeStudents;

  if (flaggedGroups.length === 0) {
    recommendations.push(
      `All ${totalStudents} students are in groups of ${k}+. No re-identification risk detected from quasi-identifiers.`
    );
  } else {
    recommendations.push(
      `${safeStudents} of ${totalStudents} students are in groups of ${k}+. ` +
        `${flaggedCount} student${flaggedCount === 1 ? ' has' : 's have'} attribute combinations shared by fewer than ${k} others.`
    );

    for (const group of flaggedGroups) {
      const attrDesc = Object.entries(group.attributes)
        .map(([key, val]) => `${key}=${val}`)
        .join(', ');
      recommendations.push(
        `Group (${attrDesc}): ${group.count} student${group.count === 1 ? '' : 's'} — consider generalising or suppressing distinguishing attributes.`
      );
    }
  }

  if (rareCategories.length > 0) {
    for (const rare of rareCategories) {
      let msg = `Rare category in "${rare.columnName}": "${rare.value}" appears only ${rare.count} time${rare.count === 1 ? '' : 's'}.`;
      if (rare.suggestion) {
        msg += ` Consider generalising to "${rare.suggestion}".`;
      }
      recommendations.push(msg);
    }
  }

  return recommendations;
}

File: packages/shield-py/shield/risk.py
"""
Re-identification risk analysis: k-anonymity and rare-category detection.
Pure functions, stdlib only.
"""

import json
from pathlib import Path
from typing import Any, Dict, List, Optional, Union

# Load generalisations
_DATA_DIR = Path(__file__).parent / 'data'
with open(_DATA_DIR / 'generalisations.json', 'r', encoding='utf-8') as _f:
    _generalisations = json.load(_f)


def analyse_risk(
    data: List[List[Any]],
    config: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Analyse re-identification risk of a dataset.

    Args:
        data: Array of row arrays (no header row).
        config: Configuration with quasi_identifiers, categorical_columns, k_threshold.

    Returns:
        Dict with k_anonymity, rare_categories, recommendations.
    """
    if config is None:
        config = {}

    quasi_identifiers = config.get('quasiIdentifiers', config.get('quasi_identifiers', []))
    categorical_columns = config.get('categoricalColumns', config.get('categorical_columns', []))
    k_threshold = config.get('kThreshold', config.get('k_threshold', 'auto'))

    total_students = len(data)

    if k_threshold == 'auto':
        k = 5 if total_students < 20 else 3
    else:
        k = k_threshold

    k_anonymity = _compute_k_anonymity(data, quasi_identifiers, k, total_students)
    rare_categories = _detect_rare_categories(data, categorical_columns, k)
    recommendations = _build_recommendations(k_anonymity, rare_categories, k)

    return {
        'kAnonymity': k_anonymity,
        'rareCategories': rare_categories,
        'recommendations': recommendations,
    }


def _compute_k_anonymity(
    data: List[List[Any]],
    qi_cols: List[int],
    k: int,
    total_students: int,
) -> Dict[str, Any]:
    """Group rows by quasi-identifier values and flag small groups."""
    if total_students == 0 or len(qi_cols) == 0:
        return {
            'threshold': k,
            'totalStudents': total_students,
            'safeStudents': total_students,
            'flaggedGroups': [],
        }

    # Build groups keyed by composite quasi-identifier string
    groups: Dict[str, Dict[str, Any]] = {}

    for i, row in enumerate(data):
        key = '\x00'.join(str(row[col] if col < len(row) else '') for col in qi_cols)
        if key not in groups:
            groups[key] = {
                'attrs': [row[col] if col < len(row) else '' for col in qi_cols],
                'indices': [],
            }
        groups[key]['indices'].append(i)

    flagged_groups = []
    flagged_count = 0

    for group in groups.values():
        if len(group['indices']) < k:
            attributes = {}
            for idx, col in enumerate(qi_cols):
                attributes[f'col{col}'] = group['attrs'][idx]
            flagged_groups.append({
                'attributes': attributes,
                'count': len(group['indices']),
                'rowIndices': group['indices'],
            })
            flagged_count += len(group['indices'])

    # Sort by count ascending
    flagged_groups.sort(key=lambda g: g['count'])

    return {
        'threshold': k,
        'totalStudents': total_students,
        'safeStudents': total_students - flagged_count,
        'flaggedGroups': flagged_groups,
    }


def _detect_rare_categories(
    data: List[List[Any]],
    categorical_columns: List[Dict[str, Any]],
    k: int,
) -> List[Dict[str, Any]]:
    """Detect rare categories in categorical columns."""
    if not categorical_columns or len(data) == 0:
        return []

    rare_threshold = min(3, k)
    results = []

    for col_info in categorical_columns:
        index = col_info['index']
        name = col_info['name']

        # Count occurrences
        counts: Dict[Any, int] = {}
        for row in data:
            val = row[index] if index < len(row) else ''
            if val is None:
                val = ''
            counts[val] = counts.get(val, 0) + 1

        for value, count in counts.items():
            if count < rare_threshold:
                suggestion = _lookup_generalisation(name, value)
                results.append({
                    'column': index,
                    'columnName': name,
                    'value': value,
                    'count': count,
                    'suggestion': suggestion,
                })

    # Sort by count ascending, then by column index
    results.sort(key=lambda r: (r['count'], r['column']))

    return results


def _lookup_generalisation(category_name: str, value: Any) -> Optional[str]:
    """Look up a generalisation suggestion for a given category name and value."""
    str_value = str(value)

    # Try exact category match first (case-insensitive)
    for cat_key, mapping in _generalisations.items():
        if cat_key.lower() == category_name.lower():
            for map_key, suggestion in mapping.items():
                if map_key.lower() == str_value.lower():
                    return suggestion
            return None

    # Category not found — search all categories for the value
    for mapping in _generalisations.values():
        for map_key, suggestion in mapping.items():
            if map_key.lower() == str_value.lower():
                return suggestion

    return None


def _build_recommendations(
    k_anonymity: Dict[str, Any],
    rare_categories: List[Dict[str, Any]],
    k: int,
) -> List[str]:
    """Build plain-English recommendation strings."""
    recommendations = []
    total_students = k_anonymity['totalStudents']
    safe_students = k_anonymity['safeStudents']
    flagged_groups = k_anonymity['flaggedGroups']

    if total_students == 0:
        recommendations.append('No data to analyse.')
        return recommendations

    flagged_count = total_students - safe_students

    if len(flagged_groups) == 0:
        recommendations.append(
            f'All {total_students} students are in groups of {k}+. '
            f'No re-identification risk detected from quasi-identifiers.'
        )
    else:
        has_str = ' has' if flagged_count == 1 else 's have'
        recommendations.append(
            f'{safe_students} of {total_students} students are in groups of {k}+. '
            f'{flagged_count} student{has_str} attribute combinations shared by fewer than {k} others.'
        )

        for group in flagged_groups:
            attr_desc = ', '.join(
                f'{key}={val}' for key, val in group['attributes'].items()
            )
            s = '' if group['count'] == 1 else 's'
            recommendations.append(
                f'Group ({attr_desc}): {group["count"]} student{s} '
                f'— consider generalising or suppressing distinguishing attributes.'
            )

    if rare_categories:
        for rare in rare_categories:
            s = '' if rare['count'] == 1 else 's'
            msg = (
                f'Rare category in "{rare["columnName"]}": '
                f'"{rare["value"]}" appears only {rare["count"]} time{s}.'
            )
            if rare['suggestion']:
                msg += f' Consider generalising to "{rare["suggestion"]}".'
            recommendations.append(msg)

    return recommendations
